lib/ui/viewport_canvas.dart

```diff
@@ -0,0 +1,511 @@
+      ..save();
+    this.position = position;
+    _dirty = true;
+    if (_debugChunks) {
+      canvas.dataset['chunkId'] = '$_nextChunkId';
+      _nextChunkId++;
+    }
+  }
+
+  final CanvasElement canvas;
+
+  CanvasRenderingContext2D get context => _context;
+  CanvasRenderingContext2D _context;
+  bool _empty = true;
+
+  bool get dirty => _dirty;
+  bool _dirty;
+
+  bool attached = false;
+  int _lastFrameRendered = -1;
+  Rect rect;
+
+  void markNeedsPaint() {
+    _dirty = true;
+  }
+
+  void markPainted() {
+    _dirty = false;
+    _empty = false;
+  }
+
+  _ChunkPosition get position => _position;
+  _ChunkPosition _position;
+
+  set position(_ChunkPosition p) {
+    if (_position == p) return;
+    _position = p;
+    rect = Rect.fromLTWH(
+      (position.x * chunkSize).toDouble(),
+      (position.y * chunkSize).toDouble(),
+      chunkSize.toDouble(),
+      chunkSize.toDouble(),
+    );
+    canvas.style.transform = 'translate(${rect.left}px, ${rect.top}px)';
+
+    context
+      ..restore()
+      ..save()
+      // Modify the canvas's coordinate system to match the global coordinates.
+      // This allows user code to paint to the canvas without knowledge of what
+      // the coordinate system of the chunk is.
+      ..translate(-rect.left, -rect.top);
+
+    _debugPaint();
+    markNeedsPaint();
+  }
+
+  void clear() {
+    if (_empty) return;
+    _context.clearRect(rect.left, rect.top, chunkSize, chunkSize);
+    _debugPaint();
+    _empty = true;
+  }
+
+  void _debugPaint() {
+    if (_debugChunks) {
+      _context
+        ..save()
+        ..strokeStyle = 'red'
+        ..fillStyle = 'red'
+        ..fillStyle = '#DDDDDD'
+        ..fillRect(rect.left + 2, rect.top + 2, rect.width - 2, rect.height - 2)
+        ..fillText('$rect', rect.left + 50, rect.top + 50)
+        ..restore();
+    }
+  }
+}
+
+class _ChunkPosition {
+  _ChunkPosition(this.x, this.y);
+
+  @override
+  bool operator ==(dynamic other) {
+    if (other is! _ChunkPosition) return false;
+    return y == other.y && x == other.x;
+  }
+
+  @override
+  int get hashCode => y * 37 + x;
+
+  final int x;
+  final int y;
+}
+
+typedef CanvasPaintCallback = void Function(
+    CanvasRenderingContext2D context, Rect rect);
+
+typedef MouseCallback = void Function(Offset offset);
+
+/// Class that enables efficient rendering of an arbitrarily large canvas by
+/// managing a set of [chunkSize] x [chunkSize] tiles and only rendering tiles
+/// for content within the current viewport.
+///
+/// This class is only compatible with browsers that support
+/// [IntersectionObserver].
+/// https://caniuse.com/#feat=intersectionobserver
+class ViewportCanvas extends Object with SetStateMixin {
+  ViewportCanvas({
+    @required CanvasPaintCallback paintCallback,
+    MouseCallback onTap,
+    MouseCallback onMouseMove,
+    VoidCallback onMouseLeave,
+    String classes,
+    this.addBuffer = true,
+  })  : _paintCallback = paintCallback,
```

Comment:
can you use `this.paramName` for some of these?

Reply:
If these were positional parameters I would do that but for named parameters keeping the parameters free of underscores is important.